// pii-redact/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AddressMatcher: () => AddressMatcher,
  CreditCardMatcher: () => CreditCardMatcher,
  DateOfBirthMatcher: () => DateOfBirthMatcher,
  DefaultMatchers: () => DefaultMatchers,
  DriverLicenseMatcher: () => DriverLicenseMatcher,
  EmailMatcher: () => EmailMatcher,
  HashStrategy: () => HashStrategy,
  IPv4Matcher: () => IPv4Matcher,
  LocationCoordinatesMatcher: () => LocationCoordinatesMatcher,
  MaskStrategy: () => MaskStrategy,
  MentionMatcher: () => MentionMatcher,
  NlpMatcher: () => NlpMatcher,
  PassportMatcher: () => PassportMatcher,
  PhoneMatcher: () => PhoneMatcher,
  Redactor: () => Redactor,
  RegexMatcher: () => RegexMatcher,
  ReplaceStrategy: () => ReplaceStrategy,
  SSNMatcher: () => SSNMatcher,
  ZipcodeMatcher: () => ZipcodeMatcher
});
module.exports = __toCommonJS(index_exports);

// src/strategies/index.ts
var import_crypto = __toESM(require("crypto"));
var ReplaceStrategy = class {
  name = "replace";
  replacement;
  /**
   * @param replacement The string to replace the PII with. If not provided, it defaults to using the matcher name (e.g., [EMAIL]).
   */
  constructor(replacement) {
    this.replacement = replacement || "";
  }
  apply(_matchedValue, matcherName) {
    if (this.replacement) {
      return this.replacement;
    }
    return `[${(matcherName || "REDACTED").toUpperCase()}]`;
  }
};
var MaskStrategy = class {
  name = "mask";
  maskChar;
  unmaskedStart;
  unmaskedEnd;
  /**
   * @param options.maskChar The character to use for masking (default: '*')
   * @param options.unmaskedStart Number of characters to leave unmasked at the start (default: 1)
   * @param options.unmaskedEnd Number of characters to leave unmasked at the end (default: 1)
   */
  constructor(options) {
    this.maskChar = options?.maskChar || "*";
    this.unmaskedStart = options?.unmaskedStart ?? 1;
    this.unmaskedEnd = options?.unmaskedEnd ?? 1;
  }
  apply(matchedValue) {
    const length = matchedValue.length;
    if (length <= this.unmaskedStart + this.unmaskedEnd) {
      return this.maskChar.repeat(length);
    }
    const start = matchedValue.slice(0, this.unmaskedStart);
    const end = matchedValue.slice(length - this.unmaskedEnd);
    const middle = this.maskChar.repeat(length - this.unmaskedStart - this.unmaskedEnd);
    return `${start}${middle}${end}`;
  }
};
var HashStrategy = class {
  name = "hash";
  algorithm;
  /**
   * @param algorithm The cryptographic hash algorithm to use (default: 'sha256')
   */
  constructor(algorithm = "sha256") {
    this.algorithm = algorithm;
  }
  apply(matchedValue) {
    return import_crypto.default.createHash(this.algorithm).update(matchedValue).digest("hex");
  }
};

// src/redactor.ts
var import_node_html_parser = require("node-html-parser");
var Redactor = class {
  matchers;
  defaultStrategy;
  constructor(config) {
    this.matchers = config?.matchers || [];
    this.defaultStrategy = config?.defaultStrategy || new ReplaceStrategy();
  }
  /**
   * Redacts PII from an unstructured text string.
   */
  redact(text) {
    if (!text || typeof text !== "string") return text;
    const allMatches = [];
    for (const matcher of this.matchers) {
      const results = matcher.match(text);
      for (const result of results) {
        const replacement = this.defaultStrategy.apply(result.value, result.name || matcher.name);
        allMatches.push({ start: result.start, end: result.end, replacement });
      }
    }
    if (allMatches.length === 0) return text;
    allMatches.sort((a, b) => a.start - b.start);
    let resultText = text;
    const validMatches = [];
    let lastEnd = -1;
    for (const match of allMatches) {
      if (match.start >= lastEnd) {
        validMatches.push(match);
        lastEnd = match.end;
      }
    }
    for (let i = validMatches.length - 1; i >= 0; i--) {
      const { start, end, replacement } = validMatches[i];
      resultText = resultText.substring(0, start) + replacement + resultText.substring(end);
    }
    return resultText;
  }
  /**
   * Recursively traverses an object or array and redacts any string values found.
   * 
   * @param obj The payload to redact.
   * @param options Configuration for targeting or ignoring specific object keys.
   * @param currentKey Internal parameter used for tracking the recursive JSON tree.
   */
  redactObject(obj, options, currentKey) {
    if (obj === null || obj === void 0) return obj;
    if (typeof obj === "string") {
      if (options?.keysToRedact) {
        if (!currentKey || !options.keysToRedact.includes(currentKey)) {
          return obj;
        }
        const redacted = this.redact(obj);
        if (redacted === obj && currentKey) {
          return this.defaultStrategy.apply(obj, currentKey.toUpperCase());
        }
        return redacted;
      }
      return this.redact(obj);
    }
    if (Array.isArray(obj)) {
      return obj.map((item) => this.redactObject(item, options, currentKey));
    }
    if (typeof obj === "object") {
      const result = {};
      for (const [key, value] of Object.entries(obj)) {
        if (options?.ignoreKeys?.includes(key)) {
          result[key] = value;
          continue;
        }
        result[key] = this.redactObject(value, options, key);
      }
      return result;
    }
    return obj;
  }
  /**
   * Safely redacts PII from an HTML string without corrupting tags or attributes.
   */
  redactHtml(htmlString) {
    if (!htmlString || typeof htmlString !== "string") return htmlString;
    const root = (0, import_node_html_parser.parse)(htmlString);
    const traverse = (node) => {
      if (node.nodeType === 3) {
        const textNode = node;
        if (textNode.rawText && textNode.rawText.trim()) {
          textNode.rawText = this.redact(textNode.rawText);
        }
      } else if (node.nodeType === 1) {
        const element = node;
        if (element.tagName && ["SCRIPT", "STYLE"].includes(element.tagName.toUpperCase())) {
          return;
        }
        for (const child of element.childNodes) {
          traverse(child);
        }
      }
    };
    traverse(root);
    return root.toString();
  }
};

// src/matchers/nlp.ts
var import_compromise = __toESM(require("compromise"));
var NlpMatcher = class {
  name;
  config;
  constructor(config) {
    this.name = "nlp";
    this.config = {
      detectPeople: true,
      detectOrganizations: true,
      detectLocations: true,
      ...config
    };
  }
  match(text) {
    const results = [];
    const doc = (0, import_compromise.default)(text);
    if (this.config.detectPeople) {
      const people = doc.people().out("offset");
      people.forEach((p) => {
        results.push({
          value: p.text,
          // e.g. "John Smith"
          name: "person",
          // Override name for specific replacement (e.g. [PERSON])
          start: p.offset.start,
          end: p.offset.start + p.offset.length
        });
      });
    }
    if (this.config.detectOrganizations) {
      const orgs = doc.organizations().out("offset");
      orgs.forEach((o) => {
        results.push({
          value: o.text,
          // e.g. "Google"
          name: "org",
          // Override name for specific replacement (e.g. [ORG])
          start: o.offset.start,
          end: o.offset.start + o.offset.length
        });
      });
    }
    if (this.config.detectLocations) {
      const places = doc.places().out("offset");
      places.forEach((p) => {
        results.push({
          value: p.text,
          // e.g. "New York"
          name: "location",
          // Override name for specific replacement (e.g. [LOCATION])
          start: p.offset.start,
          end: p.offset.start + p.offset.length
        });
      });
    }
    return results;
  }
};

// src/matchers/index.ts
var RegexMatcher = class {
  name;
  regex;
  constructor(name, regex) {
    this.name = name;
    if (!regex.flags.includes("g")) {
      this.regex = new RegExp(regex.source, regex.flags + "g");
    } else {
      this.regex = regex;
    }
  }
  match(text) {
    const results = [];
    let match;
    this.regex.lastIndex = 0;
    while ((match = this.regex.exec(text)) !== null) {
      results.push({
        value: match[0],
        start: match.index,
        end: match.index + match[0].length
      });
    }
    return results;
  }
};
var EmailMatcher = new RegexMatcher(
  "email",
  /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g
);
var PhoneMatcher = new RegexMatcher(
  "phone",
  /(?<!\d)(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}(?!\d)/g
);
var CreditCardMatcher = new RegexMatcher(
  "credit_card",
  /\b(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|6(?:011|5[0-9]{2})[0-9]{12}|(?:2131|1800|35\d{3})\d{11})\b/g
);
var SSNMatcher = new RegexMatcher(
  "ssn",
  /\b(?!(000|666|9))\d{3}-(?!00)\d{2}-(?!0000)\d{4}\b/g
);
var IPv4Matcher = new RegexMatcher(
  "ipv4",
  /\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b/g
);
var ZipcodeMatcher = new RegexMatcher(
  "zipcode",
  /\b\d{5}(?:[-\s]\d{4})?\b/g
);
var AddressMatcher = new RegexMatcher(
  "address",
  /\b\d{1,5}\s(?:[a-zA-Z0-9\s.,-]{1,50})\s(?:St|Street|Rd|Road|Ave|Avenue|Blvd|Boulevard|Ln|Lane|Dr|Drive|Ct|Court)\b/gi
);
var LocationCoordinatesMatcher = new RegexMatcher(
  "location",
  /[-+]?([1-8]?\d(\.\d+)?|90(\.0+)?),\s*[-+]?(180(\.0+)?|((1[0-7]\d)|([1-9]?\d))(\.\d+)?)/g
);
var PassportMatcher = new RegexMatcher(
  "passport",
  /\b(?:[A-Z]{1,2}\d{6,8}|\d{9})\b/g
);
var MentionMatcher = new RegexMatcher(
  "mention",
  /(?<=^|\s)@[a-zA-Z0-9_\-]+/g
);
var DriverLicenseMatcher = new RegexMatcher(
  "driver_license",
  // Varies heavily by state. Covers alphanumeric formats of 7-14 chars commonly seen in the US.
  // Using positive lookahead (?=.*[0-9]) to ensure it has at least one number, preventing normal 7-14 letter words from matching.
  /\b(?=.*[0-9])[A-Za-z0-9]{7,14}\b/g
);
var DateOfBirthMatcher = new RegexMatcher(
  "date_of_birth",
  // Matches common DOB formats: MM/DD/YYYY, DD/MM/YYYY, YYYY-MM-DD, MM-DD-YYYY, etc.
  /\b(?:(?:0?[1-9]|1[0-2])[-/](?:0?[1-9]|[12][0-9]|3[01])[-/](?:19|20)\d{2}|(?:19|20)\d{2}[-/](?:0?[1-9]|1[0-2])[-/](?:0?[1-9]|[12][0-9]|3[01]))\b/g
);

// src/index.ts
var DefaultMatchers = [
  EmailMatcher,
  PhoneMatcher,
  CreditCardMatcher,
  SSNMatcher,
  IPv4Matcher,
  ZipcodeMatcher,
  LocationCoordinatesMatcher,
  PassportMatcher,
  DriverLicenseMatcher,
  MentionMatcher,
  DateOfBirthMatcher,
  // Address matcher is typically noisy without NLP, so it's included but should be used with caution
  AddressMatcher
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AddressMatcher,
  CreditCardMatcher,
  DateOfBirthMatcher,
  DefaultMatchers,
  DriverLicenseMatcher,
  EmailMatcher,
  HashStrategy,
  IPv4Matcher,
  LocationCoordinatesMatcher,
  MaskStrategy,
  MentionMatcher,
  NlpMatcher,
  PassportMatcher,
  PhoneMatcher,
  Redactor,
  RegexMatcher,
  ReplaceStrategy,
  SSNMatcher,
  ZipcodeMatcher
});
